Use more smart pointers in the DOM

cdumez opened the pull request:
#### 0f2b0e3c5b5673d09a724dbad50a1ab5239cef44
<pre>
Use more smart pointers in the DOM
<a href="https://bugs.webkit.org/show_bug.cgi?id=261910">https://bugs.webkit.org/show_bug.cgi?id=261910</a>

Reviewed by Brent Fulgham.

* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
(WebCore::ComputedStylePropertyMapReadOnly::entries const):
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::removeAllChildrenWithScriptAssertion):
(WebCore::ContainerNode::appendChildCommon):
(WebCore::ContainerNode::prepend):
* Source/WebCore/dom/ContainerNode.h:
(WebCore::ContainerNode::protectedFirstChild const):
(WebCore::ContainerNode::protectedLastChild const):
* Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
(WebCore::ContentVisibilityDocumentState::observe):
(WebCore::ContentVisibilityDocumentState::unobserve):
* Source/WebCore/dom/DataTransferItemList.cpp:
(WebCore::DataTransferItemList::ensureItems const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::createDocumentFragment):
(WebCore::Document::setTitle):
(WebCore::Document::isSecureContext const):
* Source/WebCore/dom/Document.h:
(WebCore::Document::logger const):
* Source/WebCore/dom/DocumentInlines.h:
(WebCore::Node::protectedDocument const):
* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::DocumentMarkerController::forEachOfTypes):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::detachAttribute):
(WebCore::layoutOverflowRectContainsAllDescendants):
(WebCore::Element::insertedIntoAncestor):
(WebCore::Element::setAttributeNode):
(WebCore::Element::setAttributeNodeNS):
(WebCore::Element::updateId):
(WebCore::Element::willModifyAttribute):
(WebCore::Element::didAddAttribute):
(WebCore::Element::didModifyAttribute):
(WebCore::Element::didRemoveAttribute):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::cancelFullscreen):
(WebCore::FullscreenManager::exitFullscreen):
(WebCore::FullscreenManager::isFullscreenEnabled const):
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::treeScopeForSVGReferences const):
(WebCore::Node::defaultEventHandler):
* Source/WebCore/dom/Node.h:
(WebCore::Node::previousSibling const):
(WebCore::Node::protectedPreviousSibling const):
(WebCore::Node::nextSibling const):
(WebCore::Node::protectedNextSibling const):
(WebCore::Node::parentNode const):
* Source/WebCore/dom/NodeRareDataInlines.h:
(WebCore::NodeListsNodeData::removeCachedCollection):
* Source/WebCore/dom/Position.cpp:
(WebCore::Position::offsetForPositionAfterAnchor const):
(WebCore::Position::parentAnchoredEquivalent const):
(WebCore::Position::atFirstEditingPositionForNode const):
(WebCore::Position::atLastEditingPositionForNode const):
(WebCore::Position::atStartOfTree const):
(WebCore::Position::atEndOfTree const):
* Source/WebCore/dom/Position.h:
(WebCore::Position::protectedAnchorNode const):
* Source/WebCore/dom/PositionIterator.cpp:
(WebCore::PositionIterator::operator Position const):
(WebCore::PositionIterator::increment):
(WebCore::PositionIterator::decrement):
(WebCore::PositionIterator::atEnd const):
(WebCore::PositionIterator::atEndOfNode const):
(WebCore::PositionIterator::isCandidate const):
* Source/WebCore/dom/PositionIterator.h:
(WebCore::PositionIterator::protectedNode const):
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::processContents):
(WebCore::boundaryNodeWillBeRemoved):
* Source/WebCore/dom/RangeBoundaryPoint.h:
(WebCore::RangeBoundaryPoint::protectedContainer const):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::ScriptElement):
(WebCore::ScriptElement::childrenChanged):
(WebCore::ScriptElement::dispatchErrorEvent):
(WebCore::ScriptElement::determineScriptType const):
(WebCore::ScriptElement::prepareScript):
(WebCore::ScriptElement::requestClassicScript):
(WebCore::ScriptElement::requestModuleScript):
(WebCore::ScriptElement::requestImportMap):
(WebCore::ScriptElement::executeClassicScript):
(WebCore::ScriptElement::registerImportMap):
(WebCore::ScriptElement::executeModuleScript):
(WebCore::ScriptElement::executeScriptAndDispatchEvent):
(WebCore::ScriptElement::executePendingScript):
(WebCore::ScriptElement::ignoresLoadRequest const):
(WebCore::ScriptElement::scriptContent const):
(WebCore::ScriptElement::ref):
(WebCore::ScriptElement::deref):
* Source/WebCore/dom/ScriptElement.h:
(WebCore::ScriptElement::element):
(WebCore::ScriptElement::element const):
(WebCore::ScriptElement::protectedElement const):
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::setInnerHTML):
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/dom/SimpleRange.cpp:
(WebCore::IntersectingNodeIterator::advanceSkippingChildren):
* Source/WebCore/dom/SimpleRange.h:
(WebCore::IntersectingNodeIterator::protectedNode const):
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::nodeFromPoint):
(WebCore::TreeScope::elementsFromPoint):
(WebCore::TreeScope::focusedElementInScope):
(WebCore::TreeScope::protectedDocumentScope const):
* Source/WebCore/dom/TreeScope.h:
* Source/WebCore/html/HTMLCollection.h:
(WebCore::HTMLCollection::protectedOwnerNode const):
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::protectedElement const):

Canonical link: <a href="https://commits.webkit.org/268329@main">https://commits.webkit.org/268329@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4296743ba83e8d7f17abdcd28bb759ac611acfd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19354 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19772 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20371 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21244 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18107 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19589 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23040 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19897 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19737 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19571 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19616 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16827 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22113 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16802 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17615 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23943 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17864 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17790 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21910 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18380 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15562 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17515 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/4628 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21874 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2371 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18210 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
```diff
@@ -88,20 +88,21 @@ Vector<StylePropertyMapReadOnly::StylePropertyMapEntry> ComputedStylePropertyMap
     if (!style)
         return values;
 
+    Ref document = m_element->document();
```
Should we add a `protectedDocument()` accessor on the `m_element` class to make this a little clearer?

brentfulgham commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -1080,7 +1080,7 @@ ExceptionOr<void> ContainerNode::prepend(FixedVector<NodeOrString>&& vector)
         return { };
 
-    return insertBefore(*node, firstChild());
+    return insertBefore(*node, protectedFirstChild().get());
```
I wonder if `ContainerNode::insertBefore` should accept Refs instead of bare pointers. It Ref's the second argument in many cases inside that method.

brentfulgham commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -69,14 +69,16 @@ class ContentVisibilityIntersectionObserverCallback final : public IntersectionO
 void ContentVisibilityDocumentState::unobserve(Element& element)
 {
-    auto& state = element.document().contentVisibilityDocumentState();
+    Ref document = element.document();
```
Should we be Ref'ing Element itself? I'm not sure if `unobserve` or `removeViewportProximity` could trigger style/layout or not.

cdumez replied:
I think the idea with our new pattern is that the caller of ContentVisibilityDocumentState::unobserve() is supposed to hold a local Ref<> to the `element` it is passing as parameter.

brentfulgham commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -2553,8 +2553,10 @@ void Node::defaultEventHandler(Event& event)
-                frame->eventHandler().defaultTouchEventHandler(*renderer->node(), downcast<TouchEvent>(event));
+            if (auto* frame = document().frame()) {
+                RefPtr rendererNode = renderer->node();
+                frame->eventHandler().defaultTouchEventHandler(*rendererNode, downcast<TouchEvent>(event));
```
Should we keep the frame object alive during the touch event handler? Seems possible for script to run in that case?

brentfulgham commented on Source/WebCore/dom/Node.h:
```diff
@@ -742,10 +745,10 @@ class Node : public EventTarget {
-    Node* m_previous { nullptr };
-    Node* m_next { nullptr };
+    CheckedPtr<ContainerNode> m_parentNode;
+    TreeScope* m_treeScope;
```
Why not continue to initialize `m_treeScope` to `nullptr`?

cdumez replied:
My bad. I had converted it to a smart pointer but had to revert to a raw pointer for now. Forgot to restore the default initializer. Will fix.

brentfulgham reviewed (approved):
Looks good. r=me

cdumez committed:
Use more smart pointers in the DOM
https://bugs.webkit.org/show_bug.cgi?id=261910

Reviewed by Brent Fulgham.

* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
(WebCore::ComputedStylePropertyMapReadOnly::entries const):
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::removeAllChildrenWithScriptAssertion):
(WebCore::ContainerNode::appendChildCommon):
(WebCore::ContainerNode::prepend):
* Source/WebCore/dom/ContainerNode.h:
(WebCore::ContainerNode::protectedFirstChild const):
(WebCore::ContainerNode::protectedLastChild const):
* Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
(WebCore::ContentVisibilityDocumentState::observe):
(WebCore::ContentVisibilityDocumentState::unobserve):
* Source/WebCore/dom/DataTransferItemList.cpp:
(WebCore::DataTransferItemList::ensureItems const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::createDocumentFragment):
(WebCore::Document::setTitle):
(WebCore::Document::isSecureContext const):
* Source/WebCore/dom/Document.h:
(WebCore::Document::logger const):
* Source/WebCore/dom/DocumentInlines.h:
(WebCore::Node::protectedDocument const):
* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::DocumentMarkerController::forEachOfTypes):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::detachAttribute):
(WebCore::layoutOverflowRectContainsAllDescendants):
(WebCore::Element::insertedIntoAncestor):
(WebCore::Element::setAttributeNode):
(WebCore::Element::setAttributeNodeNS):
(WebCore::Element::updateId):
(WebCore::Element::willModifyAttribute):
(WebCore::Element::didAddAttribute):
(WebCore::Element::didModifyAttribute):
(WebCore::Element::didRemoveAttribute):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::cancelFullscreen):
(WebCore::FullscreenManager::exitFullscreen):
(WebCore::FullscreenManager::isFullscreenEnabled const):
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::treeScopeForSVGReferences const):
(WebCore::Node::defaultEventHandler):
* Source/WebCore/dom/Node.h:
(WebCore::Node::previousSibling const):
(WebCore::Node::protectedPreviousSibling const):
(WebCore::Node::nextSibling const):
(WebCore::Node::protectedNextSibling const):
(WebCore::Node::parentNode const):
* Source/WebCore/dom/NodeRareDataInlines.h:
(WebCore::NodeListsNodeData::removeCachedCollection):
* Source/WebCore/dom/Position.cpp:
(WebCore::Position::offsetForPositionAfterAnchor const):
(WebCore::Position::parentAnchoredEquivalent const):
(WebCore::Position::atFirstEditingPositionForNode const):
(WebCore::Position::atLastEditingPositionForNode const):
(WebCore::Position::atStartOfTree const):
(WebCore::Position::atEndOfTree const):
* Source/WebCore/dom/Position.h:
(WebCore::Position::protectedAnchorNode const):
* Source/WebCore/dom/PositionIterator.cpp:
(WebCore::PositionIterator::operator Position const):
(WebCore::PositionIterator::increment):
(WebCore::PositionIterator::decrement):
(WebCore::PositionIterator::atEnd const):
(WebCore::PositionIterator::atEndOfNode const):
(WebCore::PositionIterator::isCandidate const):
* Source/WebCore/dom/PositionIterator.h:
(WebCore::PositionIterator::protectedNode const):
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::processContents):
(WebCore::boundaryNodeWillBeRemoved):
* Source/WebCore/dom/RangeBoundaryPoint.h:
(WebCore::RangeBoundaryPoint::protectedContainer const):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::ScriptElement):
(WebCore::ScriptElement::childrenChanged):
(WebCore::ScriptElement::dispatchErrorEvent):
(WebCore::ScriptElement::determineScriptType const):
(WebCore::ScriptElement::prepareScript):
(WebCore::ScriptElement::requestClassicScript):
(WebCore::ScriptElement::requestModuleScript):
(WebCore::ScriptElement::requestImportMap):
(WebCore::ScriptElement::executeClassicScript):
(WebCore::ScriptElement::registerImportMap):
(WebCore::ScriptElement::executeModuleScript):
(WebCore::ScriptElement::executeScriptAndDispatchEvent):
(WebCore::ScriptElement::executePendingScript):
(WebCore::ScriptElement::ignoresLoadRequest const):
(WebCore::ScriptElement::scriptContent const):
(WebCore::ScriptElement::ref):
(WebCore::ScriptElement::deref):
* Source/WebCore/dom/ScriptElement.h:
(WebCore::ScriptElement::element):
(WebCore::ScriptElement::element const):
(WebCore::ScriptElement::protectedElement const):
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::setInnerHTML):
* Source/WebCore/dom/ShadowRoot.h:
* Source/WebCore/dom/SimpleRange.cpp:
(WebCore::IntersectingNodeIterator::advanceSkippingChildren):
* Source/WebCore/dom/SimpleRange.h:
(WebCore::IntersectingNodeIterator::protectedNode const):
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::nodeFromPoint):
(WebCore::TreeScope::elementsFromPoint):
(WebCore::TreeScope::focusedElementInScope):
(WebCore::TreeScope::protectedDocumentScope const):
* Source/WebCore/dom/TreeScope.h:
* Source/WebCore/html/HTMLCollection.h:
(WebCore::HTMLCollection::protectedOwnerNode const):
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::protectedElement const):

Canonical link: https://commits.webkit.org/268329@main
